### ISSUE TITLE
Evaluate code for univariate forecasting task maybe there is information leaking. 

### ISSUE THREAD

**Issue by wiwi** [yuezhihan/ts2vec#36]
![image](https://github.com/yuezhihan/ts2vec/assets/108532/c8182f81-5837-4d91-a045-3b0452e011d4)

Firstly, I'm sure it wasn't intentional. I reopen this issue and just want to discuss this question clearly. Maybe I'm wrong, but after I read the code, especially encode() detailed, I still cannot understand the following words:

> For example, we assume t is the first timestamp in the test set. For the first sample in the test set, the original input is [t-padding, t]. However, your input is [t, t], which feeds only one timestamp as input, resulting in poor performance and biased distribution.

First, the whole series is fed into encode() in the  eval_forecasting().  Does encode() know where the test part starts? In fact, there is no related variable t in the eval_forecasting() to indicate encode() the series is the training set or test set. I print the sliced representation of train/valid/test parts and I confirm that the encoded train/valid/test sets are overlapped. (I hope you can observe it and then reply to me.) When you encode the whole series to representation through the sliding window and sliding_length is 1. If encode() does not know where the test set starts, then the last sample of validation would cover the first test sample. But, I think this bug does not belong to encode(). 

I use Electricity as an example to explain this bug. I performed the univariate forecasting task and collected the shape of some variables. The shape of the original series(variable data in the eval_forecasting()) is (26304, 1). The shape of all_repr which is the encoded series by encode() is  (26304, 128). The shape of the representation of train/valid/test sets are (15782, 128), (5261, 128), (5261, 128). If there are no overlapped samples among the train/valid/test sets, the sum of their samples should not equal the total samples.

Of course, if you skip the head-overlapped samples, you can avoid this leakage. The generate_pred_samples() is used to discard header samples, this problem may be solved by setting reasonable parameters. But, for the valid/test sets, the parameter of the drop is set as ZERO. That means that the valid set and test set would continue to the previous section.

```
    for pred_len in pred_lens:
        train_features, train_labels = generate_pred_samples(train_repr, train_data, pred_len, drop=padding)
        valid_features, valid_labels = generate_pred_samples(valid_repr, valid_data, pred_len)
        test_features, test_labels = generate_pred_samples(test_repr, test_data, pred_len)
```

The default value of drop in the generate_pred_samples() is 0. Is the [t-padding, t] you mentioned? Are you forget to set the drop?

At last, I hope you can provide detailed evidence to prove you avoid this leakage, rather than simply asking me to read the code because I still think the bug exists after reading it again.

**Comment by yuezhihan** [comment]
1. As a common sense of GitHub, for any bug report, you should prove the existence of bug, not let the author prove that bug does not exist. As a software user, you are not required to understand the specific principles of the code, all you need to know is that the developer guarantees the following behavior:
* *Behavior 1:* In our implementation, assuming t denotes the first timestamp in test set, the input of the first sample in test set is [t-padding, t], and the input of the last sample in valid set is [t-padding-1, t-1]. Only the information in "the input" is used to generate the predicted values.

**Do you think the software implements the above behavior?**

2. This behavior can be clearly inferred from our documentation:
```py
# Sliding inference for test set
test_repr = model.encode(
    test_data,
    casual=True,
    sliding_length=1,
    sliding_padding=50
)  # n_instances x n_timestamps x output_dims
# (The timestamp t's representation vector is computed using the observations located in [t-50, t])
```
See "(The timestamp t's representation vector is computed using the observations located in [t-50, t])".

By reading the doc or code for model.encode(), you can confirm that model.encode() conforms to the behavior shown in the corrected figure:

![241101306-8693fdd2-266f-49bd-8886-d55bcc95979d](https://github.com/yuezhihan/ts2vec/assets/18026801/17060d04-7497-468c-b91b-544a7e8e83b6)

If you confirm that the figure is correct, then there is no need for you to read the code of model.encode(). This figure serves as a core point of your argument, but when it's corrected, **can you respond directly to whether this correction is right or not?**  Otherwise, I will assume you don't understand this figure and need to read the code of model.encode().

3. You mentioned overlap. Assuming t is the first timestamp in the test set, **do you think that the existence of [t-padding, t] in the test set is the data leakage you mentioned?** When we stand at timestamp t, the information from [t-padding, t-1] is known to us. Also, there is no input beyond the timestamp t-1 for valid set. Therefore, this "overlap" is not a data leakage.

I hope you can reply to **the general questions in bold type** directly with yes/no so that I know where your questions lie.

**Comment by wiwi** [comment]
1）I think you're overly emotional to write this:

> As a common sense of GitHub, for any bug report, you should prove the existence of bug, not let the author prove that bug does not exist. As a software user, you are not required to understand the specific principles of the code, all you need to know is that the developer guarantees the following behavior:

This repository is not a common open source project, but an implementation of the corresponding published paper, as other researchers will of course combine this code with the paper to verify that the content of the paper is correct or feasible. Although your paper has been published, and I think your paper is good, your code is not right for the validation of univariate forecasting. 
Because the validation code would give the wrong performance improvement, it would be unfair to other researchers. We have identified this problem, and if we do not address it, then we violate the basic spirit of science. We're not talking about the authors, we're talking about the bug of code. I hope you can deal with this problem calmly.

2）I don't think you've been reading what I've written, and I don't think your code overlaps and leaks between X and Y. Instead, there is information sharing between training set and verification set, and between verification set and test set, which leads to information leakage. In other words, part of the test sample is used to select the optimal parameters in the verification set, which will undoubtedly make the final test result better. We updated the image to make it clearer：
![image](https://github.com/yuezhihan/ts2vec/assets/108532/9c5d2299-d234-45c3-96db-781fa5541d8a)


> You mentioned overlap. Assuming t is the first timestamp in the test set, do you think that the existence of [t-padding, t] in the test input is the overlap you mentioned? When we stand at timestamp t, the information from [t-padding, t-1] is known to us. Also, there is no information beyond the timestamp t-1 for valid set. Therefore, this "overlap" is not a data leakage.

I cannot agree with you. Although the first timestamp of the test set does not extend to the verification set, some validation samples do extend to the test set. This is helpful to find the best parameter of regressor and lead to an unfair better result. 
Therefore, for univariate forecasting task, we insist that the division of training set, verification set and test set should be completed before feature extraction. 
At last, we strongly recommend that authors take a look at Informer's code. If you used the official code of informer to collect its univariate forecasting results in your paper, you should take this question more seriously because it might lead to unfair comparison. 

Note: We state that we have had no collaboration with the Informer authors team.

**Comment by yuezhihan** [comment]
1. 
>  I think you're overly emotional to write this:
> As a common sense of GitHub, for any bug report, you should prove the existence of bug, not let the author prove that bug does not exist. As a software user, you are not required to understand the specific principles of the code, all you need to know is that the developer guarantees the following behavior:

The requirement (**prove the existence of bug**) is not an emotional term. It means that the arguments should be like a complete proof. In past discussions, you have provided two arguments: code and diagrams. I pointed out their mistakes respectively. Then you ignore your two core arguments and go on to discuss other arguments. So these two arguments (code and figure) are up in the air. Besides, you mentioned "overlap" repeatedly above, but there is no proof from overlap to information leaking in your text. This is why I made a point of view 3 to make it clearly.

For another example, `rather than simply asking me to read the code because I still think the bug exists after reading it again`
There is only one place where I mentioned the word "read": "1. Please read the code of model.encode() carefully, and you can confirm that the real behavior is our corrected figure" (Sentence 1), which means:
* Proposition 1: * the code of model.encode() => the corrected figure is true
 rather than
 *Proposition 2: * the code of model.encode() => the bug does not exist
 That figure presents the process of "data => all_repr", which corresponds to model.encode(). Sentence 1 only encourages you to confirm *Proposition 1* rather than *Proposition 2*. I hope logical problems can be avoided as far as possible in discussion so that we can focus more on key arguments such as the correct form of that figure.

2. 
> I don't think you've been reading what I've written, and I don't think your code overlaps and leaks between X and Y. Instead, there is information sharing between training set and verification set, and between verification set and test set, which leads to information leakage. In other words, part of the test sample is used to select the optimal parameters in the verification set, which will undoubtedly make the final test result better. We updated the image to make it clearer.

I cherish this discussion very much and tried to understand what you said carefully. If you “don’t think the code overlaps and leaks between X and Y, “
`encoded train/valid/test sets are *overlapped*.`
`If there are no *overlapped* samples among the train/valid/test sets, the sum of their samples should not equal the total samples.`
`if you skip the head-*overlapped* samples, you can avoid this leakage.`
What is the meaning of the overlap you mentioned above? We have to define it clearly. Or the word 'overlap' is not precise? Or does overlap mean the same thing as information sharing?

3. 
Thank you for continuing to discuss your argument (figure). The problem with this figure is "data= > all_repr", which is only related to the code of model.encode. Please correct the mistakes in this first. For example, all_repr[0] should be calculated only by data[0]. Here is the corrected figure.
![241600486-9c5d2299-d234-45c3-96db-781fa5541d8a](https://github.com/yuezhihan/ts2vec/assets/18026801/d4256395-0159-482a-a319-7c99e9a18daf)

**Comment by yuezhihan** [comment]
Here is a simple example.

Say:
* time series = 0,1,2,3,4,5,6,7,8,9
* padding = 2
* pred_l = 1
* validation from 5
* testing from 7

The eval_forecasting() works as: (also shown in the corrected figure)
```
# — train set —
0 => repr[0]        predicts    1  (x)
0,1 => repr[1]      predicts    2  (x)
0,1,2 => repr[2]    predicts    3
1,2,3 => repr[3]    predicts    4
2,3,4 => repr[4]    predicts    5  (x)
# — valid set (from 5)—
3,4,5 => repr[5]    predicts    6
4,5,6 => repr[6]    predicts    7  (x)
# — test set (from 7) —
5,6,7 => repr[7]    predicts    8
6,7,8 => repr[8]    predicts    9
7,8,9 => repr[9]    (x)
```
Note that the samples marked with (x) will be dropped by generate_pred_samples().

If you still think there is a bug, please refer to the following solutions:
* If you think the code does not run as the above example, you have no choice but to read or track the code. If not consistent, to reopen the issue, code tracking results and reproduction steps are required.
* If you think the code does run as the above example HOWEVER the above example still has a leaking bug (such as 5,6,7 and 3,4,5 have an overlap of 5), you are encouraged to ask your colleagues in related fields. This data partition is a common practice for time series or quant trading, and can unbiasedly estimate out-of-sample performance in a fair manner. I don't have enough patience to answer this very basic question further.
* If you think that there is no bug in this example, but there exists a bug in other examples, please provide that example where bug exists **in the same format as this example** and reopen the issue.

---

```py
# original code
# all_repr = model.encode(
#     data[:, :, n_covariate_cols:],
#     casual=True,
#     sliding_length=1,
#     sliding_padding=padding,
#     batch_size=128
# )
# print("all_repr",data.shape, all_repr.shape)
# train_repr = all_repr[:, train_slice]
# valid_repr = all_repr[:, valid_slice]
# test_repr = all_repr[:, test_slice]

# your code
train_data = data[:, train_slice, n_covariate_cols:]
valid_data = data[:, valid_slice, n_covariate_cols:]
test_data = data[:, test_slice, n_covariate_cols:]
print("data shape:", train_data.shape, valid_data.shape, test_data.shape)

train_repr = model.encode(
    train_data,
    casual=True,
    sliding_length=1,
    sliding_padding=padding,
    batch_size=128
)
valid_repr = model.encode(
    valid_data,
    casual=True,
    sliding_length=1,
    sliding_padding=padding,
    batch_size=128
)
test_repr = model.encode(
    test_data,
    casual=True,
    sliding_length=1,
    sliding_padding=padding,
    batch_size=128
)
```

For above example, your code works as:
```
# — train set —                              
0 => repr[0]        predicts    1  (x)
0,1 => repr[1]      predicts    2  (x)
0,1,2 => repr[2]    predicts    3   #  trained with a window size of 3
1,2,3 => repr[3]    predicts    4   #  trained with a window size of 3
2,3,4 => repr[4]    predicts    5  (x)
# — valid set (from 5)—
5 => repr[5]        predicts    6   # window size = 1  => inconsistent distribution and poor performance
5,6 => repr[6]      predicts    7  (x)
# — test set (from 7) —
7 => repr[7]        predicts    8   # window size = 1  => inconsistent distribution and poor performance
7,8 => repr[8]      predicts    9   # window size = 2  => inconsistent distribution and poor performance
7,8,9 => repr[9]    (x)
```
resulting in inconsistent distribution and poor performance.

**Comment by wiwi** [comment]
Thanks auther's reply. 

In fact, I understand your description before. But you mislead the question with a very short sequence, which I think is very loose and contrary to the facts. Both ETTh/m and Electricity datasets have very large length sequences.

However, we still believe that hard splitting, which is adopted by Informer and a lot of related work (such as FEDformer, LightTS, MICN, Reformer, TImesNet, etc..), can evaluate the performance of the algorithm more objectively. So, the difference between us is the opinion of such information sharing, and therefore we believe that your corresponding comparison with Informer MAY have been conducted on a different evaluation protocol, and if so it would be unfair. 

You closed the unresolved issue again (What are you fraid?). Ok, I don't want to discuss it any more. I think people will have their own opinions.

**Comment by yuezhihan** [comment]
This issue was opened by you, and it is your responsibility to prove that the leakage exists. Since you cannot prove the existence of a bug, why do you think it is "an unresolved issue"?

Both **Behavior 1** and **the revised figure** can *deduce* that there is no information leakage. However, even if we ask you "Do you think the software implements the above behavior?”, you still don’t answer our question with YES/NO. In this case, how can we prove to you that there is no information leakage?

In the previous issue, we corrected the mistakes in your figure, and you didn't ask any questions about this correction. Then your new revised figure introduces these problems again. If you do not even respect our words, how can you expect us to communicate with you?

**Based on the above facts and many other facts listed in my previous reply, we are unable to discuss it with you.** 

## Who misled the question?

> But you mislead the question with a very short sequence, which I think is very loose and **contrary to the facts**. Both ETTh/m and Electricity datasets have **very large length sequences**.

1. During this discussion, you have provided three arguments: 1. your figure; 2. your code; 3. “overlap". Even this short example can prove that all three of your arguments are wrong:
* Your figure: for example, all_repr[0] should be calculated only by data[0].
* Your code: resulting in inconsistent distribution and poor performance.
* "overlap": for example, the sum of the samples is not equal to the total samples.

The above mistakes exist in your arguments no matter how long the sequence is.

**If you regard the correction for your arguments as "misleading", how can we have a discussion?**


2. Although we are unable to discuss it with you, we provided you with an opportunity to prove the bug **without discussion**

whether this example is **"contrary to the facts"**
> 1. If you think **the code does not run as the above example**, you have no choice but to read or track the code. If not consistent, to reopen the issue, code tracking results and reproduction steps are required.

or there is a bug in **"very large length sequences"**
> 3. If you think that there is no bug in this example, **but there exists a bug in other examples**, please provide that example where bug exists in the same format as this example and reopen the issue.

## What are we afraid of?

> What are you fraid?

We are afraid of idiots.

Note that we don't think you are. But when you ask this very personal question, we can only reply with what we really think. **As we said before, we do not encourage discussion of the discussants themselves.** Because this has nothing to do with whether a bug exists.

Since you think these issues are unresolved, **we decided to put your issues on our home page and get more people to check these issues.**

**Comment by yuezhihan** [comment]
## Fairness

> However, we still believe that hard splitting, which is adopted by Informer and a lot of related work (such as FEDformer, LightTS, MICN, Reformer, TImesNet, etc..), can evaluate the performance of the algorithm more objectively.

What researchers need is not "we believe", but evidence and logic. **Since we can completely know the range of every sample by tracking the code, why don't we come to a clear conclusion?**

**Behaviour 2**: Assuming the first timestamp of test set is t, 
* For test set in TS2Vec: the first sample inputs [t - window + 1, t], labels [t + 1, t + pred_len];
* For test set in Informer: the first sample inputs [t - window, t - 1], labels [t, t + pred_len - 1]. From the second sample to the last sample, their inputs and labels are exactly the same as their in TS2Vec;
* For valid set in TS2Vec: the last sample inputs [t - pred_len - window, t - pred_len - 1], labels [t - pred_len, t - 1];
* For valid set in Informer: the last sample inputs [t - pred_len - window, t - pred_len - 1], labels [t - pred_len, t - 1]. From the second sample to the last sample, their inputs and labels are exactly the same as their in TS2Vec;

We have strictly checked the alignment of the data. The only difference in the dataset of Informer and TS2Vec is that: Informer has one more sample in the dataset, while only one sample is negligible compared to the whole long sequence.

According to *Behaviour 2*, we conclude that there is no data leakage in both TS2Vec and Informer. Also, TS2Vec leverages the same strict evaluation method as Informer, and there is no difference in fairness.

**If you don't think so, you only need to prove that Behaviour 2 is wrong. This can be easily checked by code tracking.** For example, if the first input is not [t-window + 1, t], what is it? Is it [t-window - 10, t - 10]? Is it [t-window - 100, t - 100]? Is it [t-window - 1000, t - 1000]? Do you get the same numbers by code tracking? You've typed so much, why don't you take the time to verify if the behavior is right or wrong?

---

The following code is what you called **hard splitting** in Informer: (from Informer's data/data_loader.py)
```py
border1s = [0,   12*30*24 - self.seq_len,   12*30*24+4*30*24 - self.seq_len]
border2s = [12*30*24,   12*30*24+4*30*24,   12*30*24+8*30*24]
border1 = border1s[self.set_type]  # self.set_type:   0 for train;   1 for valid;   2 for test
border2 = border2s[self.set_type]
self.data_y = data[border1:border2]
```
The valid set:  [ T_valid - seq_len : T_test ]; 
The test set:   [ T_test - seq_len : end ].

When splitting the original data in Informer, the data are also not completely separated firstly (has an overlap of seq_len), which needs to be **further guaranteed** by the sample generating function. In our code, this is guaranteed by the casual inference. These two different implementations achieve the same behaviour.

The difference is nothing more than the order of splitting data and generating samples (with windows). It is also very common to generate samples first and then split data. These two ways are equivalent.
